README.md
# zcash-ios-wallet

[![Build Status](https://travis-ci.org/zcash/zcash-ios-wallet.svg?branch=master)](https://travis-ci.org/zcash/zcash-ios-wallet)


# Motivation
Dogfooding - _transitive verb_ - is the practice of an organization using its own product. This app was created to help us learn.

Please take note: the wallet is not an official product by ECC, but rather a tool for learning about our libraries that it is built on. This means that we do not have robust infrastructure or user support for this application. We open sourced it as a resource to make wallet development easier for the Zcash ecosystem.

# Disclaimers
There are some known areas for improvement:

- This app is mainly intended for learning and improving the related libraries that it uses. There may be bugs.
- Traffic analysis, like in other cryptocurrency wallets, can leak some privacy of the user.
- The wallet requires a trust in the server to display accurate transaction information. 

See the [Wallet App Threat Model](https://zcash.readthedocs.io/en/latest/rtd_pages/wallet_threat_model.html)
for more information about the security and privacy limitations of the wallet.

If you'd like to sign up to help us test, reach out on discord and let us know! We're always happy to get feedback!

# Description

iOS wallet using the Zcash iOS SDK that is maintained by core developers.

This a reference wallet for the following set of features:
- z2z transactions w/ encrypted memos
- reply-to formatted memos
- z2t transactions
- transparent receive-only
- autoshielding on threshold from receive only t-address

note: z means sapling shielded addresses.
# Prerequisites
* make sure you can build ZcashLightClientKit Demo Apps successfully

# Building the App
1. Clone the project, make sure you have the latest Xcode Version

2. Create `env-vars.sh file` at `${SRCROOT}` [See Instructions](https://github.com/zcash/ZcashLightClientKit#setting-env-varsh-file-to-run-locally)

3. open the `ECC-Wallet.xcodeproj` file

4. build and run on simulator.

If you had problems in any of these steps please check the [troubleshooting document](/TROUBLESHOOTING.md)
# Enabling/Disabling logging Features

This is our internal dogfooding app, and we implemented some level of event logging to be able to study user interactions and improve the User Experience with Zcash on mobile devices.

## No logs please

*You can build and run the app **without it** by using the `wallet-no-logging`*

The "no logging" target does not even build or include the Mixpanel sdk on your build or any code related with it. You can make sure of this by looking at the code yourself. If you think there's a better way achieve this, please open an Issue with your proposal. :-) 

## I want to use Mixpanel

If you wish to use mixpanel in your own build, make sure to include the following line in your env-vars.sh file
`export MIXPANEL_TOKEN="YOUR_TOKEN"`
And build the `wallet` target. Sourcery will pick it up and generate the Constants.generated.swift file that the Mixpanel SDK will use to send the events to your board

## I use some other kind of tracker...
that I would like to include in my project. The app does not care about the details of the event logger as long as it implements this protocol
````Swift
protocol EventLogging {
    func track(_ event: LogEvent, properties: KeyValuePairs<String, String>)
}
````

You can implement your own tracker proxy

# Contributing

Contributions are very much welcomed! Please read our [Contributing Guidelines](/CONTRIBUTING.md) and [Code of Conduct](/CONDUCT.md). Our backlog has many Issues tagged with the `good first issue` label. Please fork the repo and make a pull request for us to review.

# Reporting an issue

If you wish to report a security issue, please follow our [Responsible Disclosure guidelines](https://github.com/zcash/ZcashLightClientKit/blob/master/responsible_disclosure.md).

 For other kind of inquiries, feel welcome to open an Issue if you encounter a bug or would like to request a feature.

 # License

 MIT


CHANGELOG.md
# Changelog
## 0.5.0 build 139: Adoption of 0.17.0-beta.rc1
The wallet has been updated to use [Unified Addresses](https://zips.z.cash/zip-0316)
as the primary address format. Unified addresses generated by the wallet currently
contain parts that allow the wallet to receive funds to the Sapling and transparent
pools, and makes these legacy addresses available for if you need to provide an
address to a service that doesn't yet understand the Unified Address format. 

As part of this update, you may notice that the transparent address for your wallet
has changed! The new transparent address you see in your wallet is derived from 
your wallet's unified address, but don't worry; your old address will still continue 
to receive funds and make them available for shielding as normal. 

Privacy is Normal. 





## 0.5.0 build 135
* [#313] Application Hangs after Syncing on iOS 16-beta
* [#312] Download is interrupted every 30 seconds
* [#311] Show Download and Scanning Progress as block height
* (origin/adopt_0.16.0) Adopt WalletBalance and Zatoshi. Remove ReadableBalance
* Fix: warnings on ForEach
* Remove Warning for `defaultFee()` and clean up comment and spacing
* Mark existing issue on code
* [#308] Memory leak on WalletBalanceBreakDownViewModel
* [#306] Adopt ZcashLightClientKit 0.16.3-beta
## 0.5.0 build 128
* Fix: t-address being displayed instead of Sapling
* point to SDK 0.12.0-beta.5

## 0.5.0 build 125
* Issue #293 Crash when browsing Wallet History and sync happens

## 0.5.0 build 124
* [Enhancement] Remove the need for ZCASH_NETWORK_ENVIRONMENT env var. Use targets instead
* bump to 0.12.0-beta.2
## 0.5.0 build 123
* Fix: Issue #286 Force the app to be awake while syncing
## 0.5.0 build 122
* Fix: layout problems on shield dialog. added blockheight to wallet breakdown
## 0.5.0 build 121
* Fix: Issue #285 Sync does not resume when app has been backgrounded and brought back to foreground
* Minimum target iOS 14
* [New] See Details and Done Buttons for autoshielding
## 0.4.0 build 117
* FIX: privacy-leak bug fix [(see link)](https://electriccoin.co/blog/privacy-leak-bug-discovered-in-nighthawk-and-ecc-wallets/)
##  0.4.0 build 116
* FIX: Issue #126 Download params instead of bundling them
* Fix Archive being broken
##  0.4.0 build 115
* Try to copy parameters from main bundle before entirely removing them
* update CombineUrlSessionDownloader
* Combine sapling downloaded plus tests. had to fix test schemes because they not build because of output files
* add combine downloader tests
* Remove Support for iOS 13 on logging targets
##  0.4.0 build 114
* Fix: Navigation Bar does not appear on balance breakdown screen Issue #275
##  0.4.0 build 113
 * fix build error on first build on clean repo
 * Fix rebase problem
 * Adopt Zcash SDK 0.12.0-alpha.3
## 0.3.7 build 110
* Issue #271 OhMyScreen shown after background tasks fails
## 0.3.7 build 109
* Fix Issue #269 - migration error recovery
* Issue #264
* AX Fix
* updated showing dropdown based on store and rcc teams
* Merge pull request #268 from zcash/shield-poc
* update travis.yml
* fix no-logging target compile error

## 0.3.7 build 108
* Fix Home Screen layout broken on large fonts
* Add validation errors to compactBlockProcessor
* Fix: Crash on background

## 0.3.7 build 107
* Implement wipe #263
## 0.3.7 build 106
* The UnScreen is a screen that let's you navigate to a safe place instead of crashing horribly
## 0.3.7 build 105
* fix: issue #277 crash when launching
## 0.3.7 build 104
* adopt create -> prepare API
* Make Bugsnag Great Again
* Fix: shielding screen acting weird
* Fix: issue #260 FlowError.InvalidEnvironment error when sending
* FIX: differenciate between TAZ and ZEC
## 0.3.7 build 102
* Fix #219 Biometrics Locked when user has no biometrics at all
* Use presentationMode to dismiss
* Remove Awesome Menu. fix Crash on launch
## 0.3.7 build 101
* don't spin up BG Tasks on simulator
* fix Wallet balance breakdown to highlight first n decimals
* Add target for testnet only
* add handled exception tracking to bugsnag
* add mixpanel events
* Fixed Received funds UI bug on tAddr accesory view. 
* Balance breakdown 


## 0.3.7 build 100
* Balance refactor

## 0.3.7 build 99
* fix wrong text, adjust title size, add accessory view for transparent
## 0.3.7 build 98
Receive from T Address
## 0.3.7 build 97
* Fixed: issue #241 screen says enter shielded when it can accept both


## 0.3.7 build 96
* make firstView() a ViewBuilder function
* comment useless preview
* Surface errors to UI
* Fix: error when backgrounding simulator
* Fix: don't display balance when syncing
* Fix Background Task warning when app loses focus but not foreground
* log initialization crash before crashing since appstore does not catch it
* Fix profile screen getting stuck after rescan starts


## 0.3.7 build 93
* change quick rescan to be one week of blocks
* [NEW] solicited feedback dialog
* Re scan feature
* Z->T restore

## 0.3.7 build 91
* save last used seetings and if user ever shielded
* Erase and rewind #247
* Add file logger (#244)
## 0.3.4 build 80
* Issue #239 show last used address when sending
* Issue #234 
* Issue #210
* Issue #197 Wallet History does not show up when app is offline
* wrong top padding on see details screen
* fix memory leak on reset
* Don't show network fee on received transaction
* Transaction Details as no top padding


## 0.3.4 build 75
* Wallet History Navigation reset fix
* Decouple Keypad and Home, add ZECCEnvironment as environmentValue
## 0.3.3 build 74
* send flow is cancelled when synchronizer finds a new block
* remove constants file from no-logging target
## 0.3.2 build 71\
* Fixes Issue #215 - Can't paste into the memo field
* Fixes Issue #213
* fixes Issue #220
* Fixes Issue #216 Touching payment address to copy to clipboard doesn't show confirmation that it worked
* Fixes issue #218 can't copy the memo text, including can't copy the reply-to address
* FIX: phantom seed when upgrading from old wallets

wallet/wallet/Utils/Colors.swift
//
//  Colors.swift
//  wallet
//
//  Created by Francisco Gindre on 12/30/19.
//  Copyright © 2019 Francisco Gindre. All rights reserved.
//

import Foundation
import SwiftUI

extension Color {
    /// \#FFB900 Color(red: 1, green: 185.0/255.0, blue: 0)
    static let zYellow = Color(red: 1, green: 185.0/255.0, blue: 0)
    
    static let arrrBlue = Color(red: 62.0/255.0, green: 180.0/255.0, blue: 206.0/255.0)
    
    /// Transparent things blue
    /// \#4A90E2 Color(red:24/255, red: 144/255, blue: 226/255)
    static let zTransparentBlue = Color(red: 24/255, green: 144/255, blue: 226/255)
    
    /// \#FFD000 Color(red: 1, red: 0.82, blue: 0)
    static let zAmberGradient0 = Color(red: 1.0, green: 0.82, blue: 0.0)
    /// Color(red: 1.0, green:0.64, blue:0.0)
    static let zAmberGradient1 = Color(red: 1.0, green:0.64, blue:0.0)
    
    /// \#FF9400
    static let zAmberGradient2 = Color(red: 1.0, green:0.74, blue:0.0)
    
    ///  \#FFB300 Color(red: 1.0, green:0.70, blue:0.0)
    static let zAmberGradient3 = Color(red: 1.0, green:0.70, blue:0.0)
    
    /// \#FF9300 Color(red: 1.0, green: 0.58, blue: 0.0)
    static let zAmberGradient4 = Color(red: 1.0, green: 0.58, blue: 0.0)
    
    static let zBlackGradient1 = Color(red:0.16, green:0.16, blue:0.17)
    
    static let zBlackGradient2 = Color.black
    
    static let zDarkGradient1 = Color.init(red: 0.13, green: 0.14, blue: 0.15)
    
    static let zDarkGradient2 = Color.init(red: 0.11, green: 0.12, blue: 0.14)
    
    /// \#282828 Color(red: 0.17, green: 0.17, blue: 0.17)
    static let zGray = Color(red:0.16, green:0.16, blue:0.16)
    
    static let zSlightLightGray = Color(red:31/255, green:32/255, blue:37/255)
    
    /// \#4A4A4A Color(red: 74/255, green: 74/255, blue: 74/255)
    static let zGray2 = Color(red: 74/255, green: 74/255, blue: 74/255)
    
    /// \#656565 Color(red: 0.4, green: 0.4, blue: 0.4)
    static let zGray3 = Color(red: 0.4, green: 0.4, blue: 0.4)
    
    /// Color(red: 151/255, green: 151/255, blue: 151/255)
    static let zLightGray = Color(red: 151/255, green: 151/255, blue: 151/255)
    
    /// \#D8D8D8 Color(red:0.85, green:0.85, blue:0.85)
    static let zLightGray2 = Color(red:0.85, green:0.85, blue:0.85)
    
    /// Color(red: 0.17, green: 0.17, blue: 0.17)
    static let zDarkGray1 = Color(red: 0.17, green: 0.17, blue: 0.17)
    
    /// \#171717 Color(red: 0.09, green: 0.09, blue: 0.09)
    static let zDarkGray2 = Color(red: 0.09, green: 0.09, blue: 0.09)
    
    /// \#A7A7A7 Color(red: 0.65, green: 0.65, blue: 0.65)
    static let zDarkGray3 = Color(red: 0.65, green: 0.65, blue: 0.65)
    
    /// \#F7F7F7
    static let zDudeItsAlmostWhite = Color(red: 247/255, green: 247/255, blue: 247/255)
   
    static let zTextLightGray = Color(red: 107/255, green: 110/255, blue: 118/255)
   
    
    
    /// # Card colors
    
    
    /// \#FF4CA6 red:1.00, green:0.30, blue:0.65
    static let zNegativeZecAmount = Color(red:1.0, green:0.30, blue:0.65)
    
    /// \#2AFF6E Color(red:0.16, green:1.00, blue:0.43)
    static let zPositiveZecAmount = Color(red:0.16, green:1.00, blue:0.43)
    
    
    /// # Paid Card gradient
    
    /// \#FFB322 Color(red:1.0, green:0.70, blue:0.13)
    static let zPendingCardGradient1 = Color(red:1.0, green:0.70, blue:0.13)
    
    /// \# FF4242 Color(red:1.0, green:0.26, blue:0.26)
    static let zPendingCardGradient2 = Color(red:1.0, green:0.26, blue:0.26)
    
    // \# FFD649 Color(red: 1, green: 214/255, blue: 73/255)
    static let zPaidCardGradient1 = Color(red: 1, green: 214/255, blue: 73/255)
    
    // \# FFA918 Color(red: 1, green: 169/255, blue: 24/255)
    static let zPaidCardGradient2 = Color(red: 1, green: 169/255, blue: 24/255)
    
    // \#D2D2D2 Color(red: 210/255, green: 210/255, blue: 210/255)
    static let zFailedCardGradient1 = Color(red: 210/255, green: 210/255, blue: 210/255)
    
    // \#838383 Color(red: 131/255, green: 131/255, blue: 131/255)
    static let zFailedCardGradient2 = Color(red: 131/255, green: 131/255, blue: 131/255)
    
    // \#7DFF81 Color(red:125/255, green: 1, blue: 129/255)
    static let zReceivedCardGradient1 = Color(red:125/255, green: 1, blue: 129/255)
    
    // \#42EEFF Color(red:66/255, green: 238/255, blue: 1)
    static let zReceivedCardGradient2 = Color(red:66/255, green: 238/255, blue: 1)
    
    /// #Hold Button
    /// \#979797  Color(red:0.59, green:0.59, blue:0.59)
    static let zHoldButtonGray = Color(red:0.59, green:0.59, blue:0.59)
    
    // Home tab view colors
    
    static let arrrBarTintColor = Color.init(red: 0.13, green: 0.14, blue: 0.15)
    
    static let arrrBarAccentColor = Color.init(red: 194.0/255.0, green: 136.0/255.0, blue: 101.0/255.0)
    
    
    /// Amount Breakdown Gray
    
    static let zLeastSignificantAmountGray = Color(red: 119/255, green: 119/255, blue: 119/255)
    
    static let zBalanceBreakdownGradient1 = Color(red: 42/255, green: 41/255, blue: 51/255)
    
    static let zBalanceBreakdownGradient2 = Color(red: 45/255, green: 45/255, blue: 51/255)
    
    static let zBalanceBreakdownItem0 = Color(red: 44/255, green: 44/255, blue: 52/255)
    static let zBalanceBreakdownItem1 = Color(red: 134/255, green: 134/255, blue: 134/255, opacity: 0.081)
    static let zBalanceBreakdownItem2 = Color(red: 0, green: 0, blue: 0, opacity: 0.081)
    
    
    /// Settings screen colors
    static let zSettingsSectionHeader = Color.init(red: 107.0/255.0, green: 110.0/255.0, blue: 118.0/255.0)
    static let zSettingsRowBackground = Color.init(red: 25.0/255.0, green: 28.0/255.0, blue: 29.0/255.0)
    static let onColor = Color.init(red: 12.0/255.0, green: 38.0/255.0, blue: 48.0/255.0)
    static let offColor = Color.init(red: 27.0/255.0, green: 30.0/255.0, blue: 32.0/255.0)
    static let thumbOnColor = Color.init(red: 120.0/255.0, green: 176.0/255.0, blue: 193.0/255.0)
    static let thumbOffColor = Color.init(red: 83.0/255.0, green: 94.0/255.0, blue: 97.0/255.0)
    static let darkShadow = Color.init(red: 0.06, green: 0.07, blue: 0.07)
    static let lightShadow = Color.init(red: 0.26, green: 0.27, blue: 0.3)
    static let depthRoundedRectColor = Color.init(red: 29.0/255.0, green: 32.0/255.0, blue: 34.0/255.0)
    static let textTitleColor = Color.init(red: 0.59, green: 0.61, blue: 0.63)
    
    static let screenBgColor = Color.init(red: 33.0/255.0, green: 36.0/255.0, blue: 38.0/255.0)
    
    // Wallet Screen Color
    
    static let zARRRTextColor = Color.init(red: 218.0/255.0, green: 183.0/255.0, blue: 114.0/255.0)
    
    static let zARRRSentColor = Color.init(red: 218.0/255.0, green: 183.0/255.0, blue: 114.0/255.0)
    
    static let zARRRReceivedColor = Color.init(red: 26.0/255.0, green: 135.0/255.0, blue: 158.0/255.0)
    
    static let zARRRSubtitleColor = Color.init(red: 150.0/255.0, green: 150.0/255.0, blue: 160.0/255.0)
    
    static let zARRRTextColorLightYellow = Color.init(red: 132.0/255.0, green: 124.0/255.0, blue: 115.0/255.0)
    
    static let zARRRReplySelectedColor = Color.init(red: 106.0/255.0, green: 222.0/255.0, blue: 247.0/255.0)
}

extension Gradient {
    
    static let paidCard = Gradient(colors: [Color.zPaidCardGradient1, .zPaidCardGradient2])
    
    static let failedCard = Gradient(colors: [Color.zFailedCardGradient1, .zFailedCardGradient2])
    
    static let receivedCard = Gradient(colors: [Color.zReceivedCardGradient1, .zReceivedCardGradient2])
}


extension UIColor {
    static let zLightGray = UIColor(red: 151/255, green: 151/255, blue: 151/255, alpha: 1.0)
    static let zDarkGray = UIColor(red: 0.02, green: 0.02, blue: 0.02, alpha: 1)
}
